crates/kind-parser/src/lib.rs

```diff
@@ -494,8 +499,339 @@ impl<'a> Parser<'a> {
+                param,
+                r#arrow,
+                r#tilde,
+                body,
+            })))
+        } else if self.is(TokenKind::RightArrow) {
+            let r#arrow = self.bump();
+            let body = self.parse_boxed_expr()?;
+            Ok(ExprKind::Pi(Box::new(PiExpr {
+                param,
+                r#arrow,
+                r#tilde,
+                body,
+            })))
+        } else {
+            // TODO: improve this error I guess?
+            self.unexpected()
+        }
+    }
+
+    pub fn parse_rename(&mut self) -> Result<Rename> {
+        let name = self.parse_lower_id()?;
+        let r#eq = self.expect(TokenKind::Eq)?;
+        let expr = self.parse_boxed_expr()?;
+
+        Ok(Rename(name, Equal(r#eq, expr)))
+    }
+    pub fn is_named_binding(&mut self) -> bool {
+        self.peek(0).is(TokenKind::LPar) && self.peek(1).is_lower_id() && self.peek(2).is(TokenKind::Eq)
+    }
+
+    pub fn is_binding(&mut self) -> bool {
+        self.peek(0).is(TokenKind::LPar) && self.peek(1).is_lower_id() && self.peek(2).is(TokenKind::Colon)
+    }
+
+    pub fn parse_named_binding(&mut self) -> Result<NamedBinding> {
+        if self.is_named_binding() {
+            Ok(NamedBinding::Named(self.parse_paren(Self::parse_rename)?))
+        } else {
+            Ok(NamedBinding::Expr(Box::new(self.parse_atom()?)))
+        }
+    }
+
+    pub fn parse_app_binding(&mut self) -> Result<Binding> {
+        let tilde = self.eat(TokenKind::Tilde);
+        let value = self.parse_named_binding()?;
+        Ok(Binding { tilde, value })
+    }
+
+    pub fn parse_app(&mut self, multiline: bool) -> Result<ExprKind> {
+        let fun = self.parse_atom()?;
+
+        let mut args = ThinVec::new();
+
+        while (!self.get().after_newline || multiline) && !self.get().is(TokenKind::EOF) {
+            if let Some(atom) = self.try_with(Self::parse_app_binding)? {
+                args.push(atom)
+            } else {
+                break
+            }
+        }
+
+        if args.is_empty() {
+            Ok(fun.data)
+        } else {
+            Ok(ExprKind::App(Box::new(AppExpr {
+                fun: Box::new(fun),
+                args,
+            })))
+        }
+    }
+
+    pub fn parse_arrow(&mut self) -> Result<ExprKind> {
+        let left = self.parse_item(|this| this.parse_app(false))?;
+
+        if let TokenKind::FatArrow | TokenKind::RightArrow = self.get().data {
+            self.parse_pi_or_lambda_end(Param::Expr(Box::new(left)), None)
+        } else {
+            Ok(left.data)
+        }
+    }
+
+    pub fn parse_ann(&mut self) -> Result<ExprKind> {
+        let value = self.parse_item(Self::parse_arrow)?;
+
+        if self.is(TokenKind::ColonColon) {
+            let r#colon = self.bump();
+            let typ = self.parse_item(Self::parse_ann)?;
+
+            Ok(ExprKind::Ann(Box::new(AnnExpr {
+                value: Box::new(value),
```

Comment:
A lot of `parse_item` calls have `Box::new` calls, i the function should have an auxiliary function like `parse_boxed_expr`, to call with `Box` return.